Update to allow use of Ruby 3.0

BiggerNoise opened the pull request:
Addresses #105 

I wasn't able to get the 3.0 syntax for the Publisher::publish() to work with ruby 2.6 although it does work with 2.7. I made my condition < 3 to ensure we didn't poke current users.

I had to include an XML parser for 3.0, happy to pull in a different one if you'd like.

I think there's a bigger P/R to change to named arguments throughout (instead of the options hash), but that's going to give you breaking changes. I'm happy to do the work, but let's open an issue so we can discuss logistics before I start throwing stuff over the wall.

BiggerNoise committed:
Update to be compatible with ruby 3.x

BiggerNoise committed:
Manage ruby versions

iHiD commented on propono.gemspec:
```diff
@@ -20,6 +20,7 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency "aws-sdk-sns"
   spec.add_dependency "aws-sdk-sqs"
+  spec.add_dependency "nokogiri"
```
Why do we need this?

BiggerNoise replied:
Very annoying I agree. For Ruby 3, I got error messages from the AWS libraries indicating that an XML parser was required. 

```
/Users/adavis/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/aws-sdk-core-3.125.4/lib/aws-sdk-core/xml/parser.rb:74:in `set_default_engine': Unable to find a compatible xml library. Ensure that you have installed or added to your Gemfile one of ox, oga, libxml, nokogiri or rexml (RuntimeError)
```

dougal replied:
This is actually kind of neat on the part of AWS. Their library needs an XML parser, but your project probably already has one of the common ones, so support all of the common ones.

There is no nice way to downstream that for `propono` users though, without a lot of explanation.

BiggerNoise replied:
The error message was ace. I dream of writing messages that useful to my users

BiggerNoise replied:
I can make nokogiri a dev dependency so we don't introduce a dependency that's satisfied elsewhere in the user's project.

BiggerNoise replied:
I have updated the gemspec to make nokogiri a development dependency

iHiD commented on lib/propono/services/publisher.rb:
```diff
@@ -5,8 +5,14 @@ class PublisherError < ProponoError
+    if RUBY_VERSION < '3'
+      def self.publish(*args)
+        new(*args).publish
+      end
```
I think `...` might actually be better here than `*args`, as that will allow for any kwparams in the future.

Btw, there's also [ruby2_keywords](https://github.com/ruby/ruby2_keywords) which might be useful (grep `ruby2_keywords` [here](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/) for some more info.

I don't think I have a preference though as long as it works.

BiggerNoise replied:
It's absolutely your library, so I'll follow your lead. My intention was to make the ruby 2 version the same as it ever was.

If I had my druthers, we'd have all the parameters spelled out on the line and remove all the argument cleverness and the use of the options hash. This would be a breaking change, so I think it's better reserved for another P/R that could be worked into a breaking version release.

iHiD replied:
OK, yeah, I'm happy to leave it as it is :)

iHiD commented on .github/workflows/tests.yml:
```diff
@@ -15,7 +15,7 @@ jobs:
         os:
           - ubuntu-latest
-        ruby-version: [2.6, 2.7]
+        ruby-version: [2.6, 2.7, 3.0.0-preview1]
```
Could you add `3.0.3` and 3.1.0` here pls? I think both are on GHA now, and there's differences between 3.0 and 3.1 with this stuff.

BiggerNoise replied:
OK, that's weird. You can see that calling out 3.1 didn't work when I first tried it (https://github.com/KoanHealth/propono/runs/4874203610?check_suite_focus=true), but ruby setup indicates that 3.1 is supported. 

Let me take another look at that.

BiggerNoise replied:
I have updated the tests.yml to test the correct versions (biggest deal was changing to ruby-setup@v1)

iHiD commented on test/services/publisher_test.rb:
```diff
@@ -11,8 +11,8 @@ def test_initialization
-      Publisher.expects(:new).with(aws_client, topic, message).returns(mock(publish: nil))
-      Publisher.publish(aws_client, topic, message)
+      Publisher.expects(:new).with(aws_client, propono_config, topic, message).returns(mock(publish: nil))
+      Publisher.publish(aws_client, propono_config, topic, message)
```
Meh. Sorry! :)

iHiD reviewed (changes requested):
Thanks for this! 

@dougal Could you take a look too please?

dougal reviewed (approved):
Looks fine. Actually simpler than I expected when I read #105.

BiggerNoise committed:
Make nokogiri a development dependency, so users can work with whatever XML parser they have in their project

BiggerNoise committed:
Expand list of tested rubies

BiggerNoise commented:
It's been ages since I made a P/R, so forgive any etiquette missteps. 

I'm going to resolve the remaining conversations b/c I think I have addressed the issues with the two commits I made today. Please let me know if there are any further changes you'd like to see.

iHiD commented on .github/workflows/tests.yml:
```diff
@@ -15,13 +15,13 @@ jobs:
 
       - name: Set up Ruby
-        uses: ruby/setup-ruby@a699edbce608a2c128dedad88e3b6a0e28687b3c
+        uses: ruby/setup-ruby@v1
```
Could you revert this please (or even better pin it to the latest sha: [180e1d1f1c5eadbfbab3dab6fb79ab0c07e3cecc](https://github.com/ruby/setup-ruby/commit/180e1d1f1c5eadbfbab3dab6fb79ab0c07e3cecc))

Pinning to tags is generally unsafe as someone in the future could take control of that repo, republish a different tag with the same version, and then we'd be running malicious code. Using Git shas guarantee that we're running the code that we actually intend to, which is why I use them.

BiggerNoise replied:
Done. Good point on the SHA vs version. FWIW - You'll need to bump this version as you bring newer rubies into the matrix.

iHiD commented on propono.gemspec:
```diff
@@ -24,6 +24,7 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "bundler", "~> 2.1"
   spec.add_development_dependency "rake"
   spec.add_development_dependency "mocha"
+  spec.add_development_dependency "nokogiri"
```
Still not clear on why we need this?

BiggerNoise replied:
In ruby 3.x, the AWS library needs to have an XML parser. I don't know why it doesn't require this in 2.x. What I do know is that if you leave this line out and test under 3.x, you'll get the following error:

```
/Users/adavis/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/aws-sdk-core-3.125.4/lib/aws-sdk-core/xml/parser.rb:74:in `set_default_engine': Unable to find a compatible xml library. Ensure that you have installed or added to your Gemfile one of ox, oga, libxml, nokogiri or rexml (RuntimeError)
```

I picked nokogiri because it's a known quantity that I have installed everywhere, we can pick anything. I left it as a dev dependency so the project that includes propono will resolve the dependency (because it likely already has).

iHiD commented:
@BiggerNoise It's all great, thanks. One request on the git sha and one question about nokogiri, but then I'll get it merged and cut.

BiggerNoise committed:
Pin the ruby-setup action to the latest sha (ensures that the setup we use is not changed without our knowledge)
